docs/proposals-accepted/202201-parallel-compaction.md

```diff
@@ -0,0 +1,123 @@
+
+We’ll introduce the ability for compactor to be running in planner, worker, and planner + worker mode. By running the compactor in a planner mode, the compactor will scan the object store for a list of blocks, and produce plans. The worker nodes will receive jobs from the planner nodes, and execute the compaction. Existing users do not have to spin up a new component, as they can still enjoy parallelism by vertically scaling their node.
+
+### Architecture
+
+#### Object Store Layout
+
+![block layout](../img/202201-parallel-compaction-block-layout.png)
+
+Given that we [cannot compact two blocks with the same source](https://github.com/cortexproject/cortex/pull/4433#issuecomment-1021505763), meaning that if BlockX contains BlockA, B, and BlockZ contain BlockA, C, the compaction will fail as BlockA is present in both. Hence, I propose that we put a lock file for each block that is being compacted currently. The planner will put a lock file on each block that will be sent to a worker, and a worker will update the timestamp at an interval.
+
+The planners will generate plans based on all the available blocks on object storage, but if any block in a plan has an active lock file, the plan will be dropped. If a lock file is stale, it will be considered as active.
+
+
+#### Ownership Determination
+
+![ownership determination](../img/202201-parallel-compaction-ownership-determination.png)
+A tenant will belong to only 1 single planner, while a tenant’s compaction jobs can run on multiple workers. The planners will determine the ownership of plan creation via a ring, similar to how Thanos receiver works. The planner will find out the workers via a ring, and determine which workers are available for a given tenant.
+
+#### High-level Coordination
+
+At a high level, the planner coordinates the workers, by having workers pull from the planner, and continuously generating plan while reading the lock file in blocks. Workers will participate in the ring, but they will not use the ring directly to find jobs. Planners will find the workers on the ring, to find the corresponding workers for a tenant.
+
+There are two main parts to the implementation
+
+**Planner**
+Creates plans based on the current view of the object store
+
+During startup, planner will:
+
+* discover other planners, determine the plan creation ownership amongst planners
+* check for the lock file, and create plans
+
+When a worker comes, a planner will:
+
+* determine which plan is suitable for the worker
+    * put a lock file for every block with current timestamp
+    * assign the job to a worker
+
+**Worker**
+During startup, a worker will:
+
+* talk to planners to find a job
```

Comment:
Can you please also add a section about the coordination API between planner and worker in this proposal?

Reply:
will do in the next rev.